### ISSUE TITLE
Configurable behavior for workers with a completed ciBuildId

### ISSUE THREAD

**Issue by boxofcrates** [sorry-cypress/sorry-cypress#368]
- [X] I have searched issues for similar feature requests

## What is this feature about (1 sentence)
Configurable behavior when a worker passes the director a ciBuildId corresponding to a run that has already completed. This will close a 'loophole' allowing CI pipelines to pass even when Cypress tests fail.

## Why is it needed? What is the value? For whom do we build it?
I'll explain the issue with the current behavior. I'm using Gitlab as an example, but I think this applies to other scm systems:

Setup: CI pipeline running 3 parallel Cypress workers talking to sorry-cypress. One of the specs failed, so Gitlab shows cypress-2 job as failed, but cypress-1 and cypress-3 as succeeded.
Scenario: developer sees cypress-2 job failed, clicks retry _on that job_

cypress-2 job will start with the same ciBuildId. It connects to director, is told that all specs have already been run, and successfully quits.

However... the CI pipeline now shows as succeeding, with all jobs passing. Even though the one spec failed and was never re-run. There's no way to change this behavior from the Gitlab side, as there is no env var that a) is the same across all jobs in an initial pipeline run and b) is different when a job is re-tried.

## High-level feature overview
I'm not sure what the best solution here is, but my thinking is:
 * Configurable behavior in director when a worker comes in with a ciBuildId corresponding to an already-finished job
 * Depending on configuration, either: PASS pass with no specs (current behavior), FAIL fail, or RE-RUN re-start the whole run and give all specs to the new worker

RE-RUN is IMO the ideal, but even just FAIL would allow closing this loophole.

@agoldis thoughts? If we can come up with an agree-able solution I'm happy to work on this, it'll save me from tearing my hair out in my professional life 😆

**Comment by agoldis** [comment]
@boxofcrates thanks for opening that issue! 

I tend to try using native cypress "retries" instead of complex rerun logic, see https://github.com/sorry-cypress/sorry-cypress/issues/197

So basically it's all about re-running the failed specs. Specifically, re-running *only* the failed specs. Let's try to break it down:

## Rerunning all the specs
If you manage to generate a new ciBuildId when rerunning, it will just run all the specs. cypress agent generates ciBuildId automatically for retries. See https://github.com/cypress-io/cypress/issues/8451. If you generate ciBuildId manually, you'd follow the the same path to make sure rerunning creates a new ciBuildId.

## Rerunning specific failed specs

That should be possible to implement in theory, but would be quite cumbersome. Basically we need to be able to recognize if it's a rerun and fetch specific failed specs... plus support grouping etc. I'd try avoiding this complexity. 

In general, if you use "retries" and tests still fail, and then you want another "retry" via CI, something is wrong with the tests :)

**Comment by carlos0202** [comment]
Hi guys,

Sorry for jumping into this discussion with my own request or help. In my case, I would like to be able to rerun all the specs from circleci no matter if they all passed or failed. I was trying to create a dynamic environment variable on circle-ci (holding current date-time) to add that variable to ci-build-id to produce different build ids per run with no success :'(.

Can anybody help me on this? Thanks in advance!!!

**Comment by agoldis** [comment]
@carlos0202 try using `--ci-build-id $CIRCLE_WORKFLOW_ID-$CIRCLE_BUILD_NUM`, that worked for me

**Comment by carlos0202** [comment]
> @carlos0202 try using `--ci-build-id $CIRCLE_WORKFLOW_ID-$CIRCLE_BUILD_NUM`, that worked for me

Well, this could help me but I'll be facing another issue because I have multiple cypress/run steps in my circleci config. Thanks for the help, I need to figure out how can I solve this issue.

**Comment by akociemba** [comment]
@agoldis about this 

> In general, if you use "retries" and tests still fail, and then you want another "retry" via CI, something is wrong with the tests :)

It might be an issue with infrastructure, resolved outside of the pipeline, and the user might not want to create a new pipeline, just to retry few tests (especially in the case of CI/CD deployment pipelines). In such a case, it's really useful to be able to retry only failed tests (or even the whole test suite) instead of creating a new pipeline.

A special case of the above is running tests in parallel, for example, 4 threads, and one of the threads fail. Currently retrying failed job (with default ciBuildId in case of gitlab = pipelineId) is successful without actually retrying the tests, making the pipeline false-positive. Using combined ciBuildId like pipelineId_jobId, messes with smart spec division, as in our example we have now 4 runners running all the same, undivided specs, wasting our time.

**Comment by agoldis** [comment]
@akociemba thanks, those are useful use-cases indeed. 

It's going to be quite a challenge in terms of implementation, also, it is different from the original cypress dashboard behaviour, so we'd thunk carefully about the implementation and implications.
